[Token Drop] Form

tzeweiwee opened the pull request:
# Description
Added a responsive form to create token drop

- used react-hook-form to manage state
- created a context to store data for entire token drop journey

Preview link: https://ericw-token-drop-create-form.keypom-airfoil.pages.dev/drops/token/new

## Type of Change

- [ ] Bug fix
- [x] New feature
- [ ] Refactoring
- [ ] Requires a documentation update

# Checklist:

- [ ] Code duplication check
- [ ] Positive scenario check - I have performed a self-review of my own code
- [ ] Unused code cleanup
- [ ] Clean console
- [ ] Commented code, particularly in hard-to-understand areas

# Acceptance Criteria
Validations, inputs, responsiveness, form state

# Screenshots/Screen Recording of Implementation

Desktop
![image](https://user-images.githubusercontent.com/19740800/210054435-4eaca18f-0620-4285-8db1-747647f7ce36.png)


Mobile
![image](https://user-images.githubusercontent.com/19740800/210054510-cab22ae8-d7c7-467e-a00a-ce9831383987.png)

ayungavis commented on src/common/components/Checkboxes/Checkboxes.tsx:
```diff
@@ -0,0 +1,67 @@
+  }, [value.length]);
+
+  const checkboxes = items.map((item) => {
+    // const isChecked = val.includes(item.value);
```
Could we remove the comment?

ayungavis commented on src/common/components/ConnectWalletButton/ConnectWalletButton.tsx:
```diff
@@ -14,6 +14,7 @@ export const ConnectWalletButton = ({ ...props }: ConnectWalletButtonProps) => {
       >
         Connect Wallet
       </Button>
+      {/* <Button variant="pill">ericw.near</Button> */}
```
Could we remove this comment?

ayungavis commented on src/common/components/WalletBalanceInput/CostDisplay.tsx:
```diff
@@ -0,0 +1,16 @@
+
+import { Text } from '../Typography';
+
+export const CostDisplay = ({ totalCost, balanceAmount, symbol }) => {
```
Could we describe the type of each prop? Better to avoid `any` type in the future

ayungavis commented on src/common/components/Checkboxes/Checkboxes.tsx:
```diff
@@ -0,0 +1,67 @@
+  icon?: React.ReactNode;
+}
+
+interface CheckboxesProps {
```
Could we export the interface of the component props too?

tzeweiwee replied:
May I know why we need to export a component props?

ayungavis replied:
Sometimes if needed we need to use the props for other components or files, but it's okay if we don't export it because right now it's not needed on the other

ayungavis replied:
Sorry my bad, I usually export this for creating a design system 😅, I'll resolve the conversation.

ayungavis commented on src/common/components/Checkboxes/Checkboxes.tsx:
```diff
@@ -0,0 +1,67 @@
+
+  useEffect(() => {
+    onChange(value);
+  }, [value.length]);
```
I think we need to include `onChange` to the `useEffect` dependency because we're using it inside that to ignore the ESlint warning message

```suggestion
  }, [onChange, value.length]);
```

tzeweiwee replied:
yeah onChange was previous an inline function so useEffect kept looping! but i've fixed it!

ayungavis commented on src/common/components/WalletBalanceInput/WalletSelectorMenu.tsx:
```diff
@@ -0,0 +1,77 @@
+  icon: React.ReactNode;
+}
+
+interface WalletBalanceInputProps {
```
Same as above, could we export the interface of the component props too?

tzeweiwee replied:
why? its not used anywhere else

ayungavis replied:
Resolved, the answer [here](https://github.com/keypom/keypom-airfoil/pull/17#discussion_r1059294282)

ayungavis commented on src/common/components/WalletBalanceInput/WalletSelectorMenu.tsx:
```diff
@@ -0,0 +1,77 @@
+
+import { Text } from '../Typography';
+
+export interface ITokenBalance {
+  symbol: string;
+  amount: string;
+  icon: React.ReactNode;
+}
+
+interface WalletBalanceInputProps {
+  selectedWallet: ITokenBalance;
+  tokens: ITokenBalance[];
```
Same as above, could we rename the `ITokenBalance` like the prop name because it's more understandable

```suggestion
export interface Token {
  symbol: string;
  amount: string;
  icon: React.ReactNode;
}

interface WalletBalanceInputProps {
  selectedToken: Token;
  tokens: Token[];
```

Or you can rename it to `Wallet` 🤔

tzeweiwee replied:
yeah let me renames these

ayungavis commented on src/modules/AllDrops/AllDrops.tsx:
```diff
@@ -43,6 +47,17 @@ const TABLE_DATA = [
 export default function AllDrops() {
   const { isOpen, onOpen, onClose } = useDisclosure();
+
+  const dropMenuItems = useMemo(
```
I think it's better to remove the `useMemo` because there is no dependency

ayungavis commented on src/modules/CreateDrops/TokenDrop/CreateTokenDropForm.tsx:
```diff
@@ -0,0 +1,144 @@
+    setValue('selectedFromWallet', { symbol, amount });
+  };
+
+  const handleSubmitClick = (data) => console.log(data); // temporary workaround
```
Could we add `// TODO` comment inside the handle function?

ayungavis commented on src/modules/CreateDrops/TokenDrop/CreateTokenDropForm.tsx:
```diff
@@ -0,0 +1,144 @@
+
+  const handleSubmitClick = (data) => console.log(data); // temporary workaround
+
+  console.log(errors);
```
Could we remove the console?

ayungavis commented on src/modules/CreateDrops/TokenDrop/CreateTokenDropForm.tsx:
```diff
@@ -0,0 +1,144 @@
+          control={control}
+          name="selectedToWallets"
+          render={({ fieldState: { error } }) => {
+            console.log(error);
```
Could we remove the console?

ayungavis commented on src/modules/CreateDrops/TokenDrop/CreateTokenDropForm.tsx:
```diff
@@ -0,0 +1,144 @@
+                  items={WALLET_OPTIONS}
+                  onChange={(value) => {
+                    setValue('selectedToWallets', value);
+                    console.log(value);
```
Same as above, could we remove the console?

ayungavis commented on src/modules/CreateDrops/TokenDrop/data.tsx:
```diff
@@ -0,0 +1,42 @@
+import { HereLogoIcon, MyNearLogoIcon, NearLogoIcon } from '@/common/components/Icons';
+
+/**
+ * Temporary data providers
```
Could we add `// TODO` comment in here too? Or if we will remove this file, you can ignore this

tzeweiwee replied:
we will eventually remove this file once integrated with backend so no worries

ayungavis reviewed (changes requested):
Hi @tzeweiwee, thanks for the hard work 🙌
I left some comments below. Also, I just tested the form and got some issues here:
1. The redirect link field always shows an error `Invalid url` if I have focused on that and removed my focus from the field. I think we need to validate it when the user inputs some values on that field. If the value is empty, we don't need to validate it because it's optional

https://user-images.githubusercontent.com/4987902/209962387-65f509bc-dc0d-48db-9474-af595b2d1e34.mov

2. Could we remove the `0` number in the front of the value if I have set a new value after I removed it?

https://user-images.githubusercontent.com/4987902/209962616-28cf0b26-1c4d-43d7-95f8-2b6e15eabbdc.mov

3. The redirect link field is optional. But if I have focused on that and removed my focus from the field, the validation gave me an error and make the button `disabled`

https://user-images.githubusercontent.com/4987902/209962855-ca01c9cd-bf8d-4c14-b4ef-4729fed0e1da.mov

Let me know if you have any questions about it 😉

ayungavis commented:
Hi @tzeweiwee, thanks for the update. I think no more comments from me 🙌

tzeweiwee commented:
> Hi @tzeweiwee, thanks for the hard work 🙌 I left some comments below. Also, I just tested the form and got some issues here:
> 
> 1. The redirect link field always shows an error `Invalid url` if I have focused on that and removed my focus from the field. I think we need to validate it when the user inputs some values on that field. If the value is empty, we don't need to validate it because it's optional
> 
>  Screen.Recording.2022-12-29.at.7.51.10.PM.mov 
> 2. Could we remove the `0` number in the front of the value if I have set a new value after I removed it?
> 
>  Screen.Recording.2022-12-29.at.7.52.30.PM.mov 
> 3. The redirect link field is optional. But if I have focused on that and removed my focus from the field, the validation gave me an error and make the button `disabled`
> 
>  Screen.Recording.2022-12-29.at.7.53.09.PM.mov 
> Let me know if you have any questions about it 😉

Thanks for catching these @ayungavis , i've fixed them and also added validation for wallet checkboxes
